File: metrics.py
from matrices import mat_vec, vector_sub, norm2


# невязка - проверяет, хорошо ли выполняется уравнение Ax = b
def residual_norm(a, x, b):
    return norm2(vector_sub(mat_vec(a, x), b))


# относительная погрешность - насколько найденное решение близко к точному
def relative_error(x_approx, x_exact):
    numerator = norm2(vector_sub(x_approx, x_exact))
    denominator = norm2(x_exact)
    if denominator == 0:
        return numerator
    return numerator / denominator

File: metrics.c
#include <stdlib.h>
#include "matrices.h"
#include "metrics.h"

// невязка - проверяет, хорошо ли выполняется уравнение Ax = b
double residual_norm(double **a, double *x, double *b, int n) {
    double *ax = mat_vec(a, x, n);
    double *diff = vector_sub(ax, b, n);
    double r = norm2(diff, n);
    free(ax);
    free(diff);
    return r;
}

// относительная погрешность - насколько решение близко к точному
double relative_error(double *x_approx, double *x_exact, int n) {
    double *diff = vector_sub(x_approx, x_exact, n);
    double num = norm2(diff, n);
    double den = norm2(x_exact, n);
    free(diff);
    if (den == 0.0)
        return num;
    return num / den;
}
